NetKet/Machine/conv_layer.hpp

```diff
@@ -0,0 +1,434 @@
+// Copyright 2018 The Simons Foundation, Inc. - All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NETKET_CONVLAYER_HH
+#define NETKET_CONVLAYER_HH
+
+#include <time.h>
+#include <Eigen/Dense>
+#include <algorithm>
+#include <complex>
+#include <fstream>
+#include <random>
+#include <vector>
+#include "Lookup/lookup.hpp"
+#include "Utils/all_utils.hpp"
+#include "abstract_layer.hpp"
+
+namespace netket {
+/** Convolutional layer with spin 1/2 hidden units.
+ Important: In order for this to work correctly, VectorType and MatrixType must
+ be column major.
+ */
+template <typename Activation, typename T>
+class Convolutional : public AbstractLayer<T> {
+  using VectorType = typename AbstractLayer<T>::VectorType;
+
+  using MatrixType = typename AbstractLayer<T>::MatrixType;
+  static_assert(!MatrixType::IsRowMajor, "MatrixType must be column-major");
+
+  Activation activation_;  // activation function class
+
+  bool usebias_;  // boolean to turn or off bias
+
+  int nv_;            // number of visible units in the full network
+  int in_channels_;   // number of input channels
+  int in_size_;       // input size: should be multiple of no. of sites
+  int out_channels_;  // number of output channels
+  int out_size_;      // output size: should be multiple of no. of sites
+  int npar_;          // number of parameters in layer
+  int kernelpar_;     // number of parameters in layer
```

Comment:
@kchoo1118 ok to remove it?

Reply:
I was wondering why the compiler doesn't pop up a warning for this though. Normally it should, if the variable is unused.

Reply:
> I was wondering why the compiler doesn't pop up a warning for this though. Normally it should, if the variable is unused.

Probably because it is assigned to in `Init`. My IDE did complain, though.